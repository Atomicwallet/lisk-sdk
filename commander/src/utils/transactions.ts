/*
 * LiskHQ/lisk-commander
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import {
	BaseTransaction,
	DelegateTransaction,
	MultisignatureTransaction,
	TransactionJSON,
	TransferTransaction,
	VoteTransaction,
} from '@liskhq/lisk-transactions';

import { ValidationError } from './error';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const parseTransactionString = (transactionStr: string): any => {
	try {
		// eslint-disable-next-line @typescript-eslint/no-unsafe-return
		return JSON.parse(transactionStr);
	} catch (error) {
		throw new ValidationError('Could not parse transaction JSON.');
	}
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const defaultTransactions: { readonly [key: number]: any } = {
	8: TransferTransaction,
	10: DelegateTransaction,
	11: VoteTransaction,
	12: MultisignatureTransaction,
};

export const instantiateTransaction = (data: TransactionJSON): BaseTransaction => {
	if (data.type === undefined || data.type === null) {
		throw new Error('Invalid transaction without type');
	}

	if (!Object.keys(defaultTransactions).includes(String(data.type))) {
		throw new Error(`Transaction type ${data.type} is not supported`);
	}
	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
	const TransactionClass = defaultTransactions[data.type];

	// eslint-disable-next-line @typescript-eslint/no-unsafe-return,@typescript-eslint/no-unsafe-call
	return new TransactionClass(data);
};
