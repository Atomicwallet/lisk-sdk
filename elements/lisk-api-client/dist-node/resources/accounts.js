"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var api_method_1 = require("../api_method");
var api_resource_1 = require("../api_resource");
var constants_1 = require("../constants");
var AccountsResource = (function (_super) {
    __extends(AccountsResource, _super);
    function AccountsResource(apiClient) {
        var _this = _super.call(this, apiClient) || this;
        _this.path = '/accounts';
        _this.get = api_method_1.apiMethod({
            method: constants_1.GET,
        }).bind(_this);
        _this.getMultisignatureGroups = api_method_1.apiMethod({
            method: constants_1.GET,
            path: '/{address}/multisignature_groups',
            urlParams: ['address'],
        }).bind(_this);
        _this.getMultisignatureMemberships = api_method_1.apiMethod({
            method: constants_1.GET,
            path: '/{address}/multisignature_memberships',
            urlParams: ['address'],
        }).bind(_this);
        return _this;
    }
    return AccountsResource;
}(api_resource_1.APIResource));
exports.AccountsResource = AccountsResource;
//# sourceMappingURL=accounts.js.map