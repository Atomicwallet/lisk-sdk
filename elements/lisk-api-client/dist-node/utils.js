"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toQueryString = function (obj) {
    var parts = Object.keys(obj).reduce(function (accumulator, key) { return __spread(accumulator, [
        encodeURIComponent(key) + "=" + encodeURIComponent(obj[key]),
    ]); }, []);
    return parts.join('&');
};
var urlParamRegex = /{[^}]+}/;
exports.solveURLParams = function (url, params) {
    if (params === void 0) { params = {}; }
    if (Object.keys(params).length === 0) {
        if (url.match(urlParamRegex) !== null) {
            throw new Error('URL is not completely solved');
        }
        return url;
    }
    var solvedURL = Object.keys(params).reduce(function (accumulator, key) {
        return accumulator.replace("{" + key + "}", params[key]);
    }, url);
    if (solvedURL.match(urlParamRegex) !== null) {
        throw new Error('URL is not completely solved');
    }
    return encodeURI(solvedURL);
};
//# sourceMappingURL=utils.js.map