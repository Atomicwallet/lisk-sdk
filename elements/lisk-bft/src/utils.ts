/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

'use strict';

import { validator } from '@liskhq/lisk-validator';

import { blockHeaderSchema } from './schema';
import { BlockHeader } from './types';

export const validateBlockHeader = (blockHeader: BlockHeader): boolean => {
	const errors = validator.validate(blockHeaderSchema, blockHeader);
	if (errors.length) {
		throw new Error(errors[0].message);
	}

	return true;
};
