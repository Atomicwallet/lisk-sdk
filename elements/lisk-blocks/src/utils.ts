/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import * as BigNum from '@liskhq/bignum';
import { hash } from '@liskhq/lisk-cryptography';

import {
	BlockHeader,
	BlockHeaderJSON,
	Storage,
	StorageTransaction,
} from './types';

export const loadBlocksFromLastBlockId = async (
	storage: Storage,
	lastBlockId: string,
	limit: number,
) => {
	if (!lastBlockId) {
		throw new Error('lastBlockId needs to be specified');
	}
	if (!limit) {
		throw new Error('Limit needs to be specified');
	}

	// Get height of block with supplied ID
	const [lastBlock] = await storage.entities.Block.get({ id: lastBlockId });
	if (!lastBlock) {
		throw new Error(`Invalid lastBlockId requested: ${lastBlockId}`);
	}

	const lastBlockHeight = lastBlock.height;

	// Calculate max block height for database query
	const fetchUntilHeight = lastBlockHeight + limit;

	const filter = {
		height_gt: lastBlockHeight,
		height_lte: fetchUntilHeight,
	};

	return storage.entities.Block.get(filter, {
		extended: true,
		limit,
		sort: ['height'],
	});
};

export const getIdSequence = async (
	storage: Storage,
	height: number,
	lastBlock: BlockHeader,
	genesisBlock: BlockHeader,
	numberOfDelegates: number,
) => {
	// Get IDs of first blocks of (n) last rounds, descending order
	// EXAMPLE: For height 2000000 (round 19802) we will get IDs of blocks at height: 1999902, 1999801, 1999700, 1999599, 1999498
	const rows: Array<Partial<
		BlockHeaderJSON
	>> = await storage.entities.Block.getFirstBlockIdOfLastRounds({
		height,
		numberOfRounds: 5,
		numberOfDelegates,
	});
	if (rows.length === 0) {
		throw new Error(`Failed to get id sequence for height: ${height}`);
	}

	const ids: string[] = [];

	// Add genesis block at the end if the set doesn't contain it already
	if (genesisBlock) {
		const partialGenesis = {
			id: genesisBlock.id,
			height: genesisBlock.height,
		};

		if (!rows.map(r => r.id).includes(partialGenesis.id)) {
			rows.push(partialGenesis);
		}
	}

	// Add last block at the beginning if the set doesn't contain it already
	if (lastBlock && !rows.map(r => r.id).includes(lastBlock.id)) {
		rows.unshift({
			id: lastBlock.id,
			height: lastBlock.height,
		});
	}

	// Extract blocks IDs
	rows.forEach(row => {
		// FIXME: Looks like double check
		if (!ids.includes(row.id as string)) {
			ids.push(row.id as string);
		}
	});

	return {
		firstHeight: rows[0].height,
		ids: ids.join(),
	};
};

export const addBlockProperties = (block: BlockHeaderJSON) => {
	block.totalAmount = new BigNum(block.totalAmount || 0);
	block.totalFee = new BigNum(block.totalFee || 0);
	block.reward = new BigNum(block.reward || 0);

	if (block.version === undefined) {
		block.version = 0;
	}
	if (block.numberOfTransactions === undefined) {
		block.numberOfTransactions =
			block.transactions === undefined ? 0 : block.transactions.length;
	}
	if (block.payloadLength === undefined) {
		block.payloadLength = 0;
	}
	if (block.transactions === undefined) {
		block.transactions = [];
	}

	return block;
};

export const deleteBlockProperties = (block: BlockHeader) => {
	const reducedBlock = {
		...block,
	};
	/* tslint:disable:no-delete */
	if (reducedBlock.version === 0) {
		delete reducedBlock.version;
	}
	// VerifyBlock ensures numberOfTransactions is transactions.length
	if (typeof reducedBlock.numberOfTransactions === 'number') {
		delete reducedBlock.numberOfTransactions;
	}
	if (reducedBlock.totalAmount.eq(0)) {
		delete reducedBlock.totalAmount;
	}
	if (reducedBlock.totalFee.eq(0)) {
		delete reducedBlock.totalFee;
	}
	if (reducedBlock.payloadLength === 0) {
		delete reducedBlock.payloadLength;
	}
	if (reducedBlock.reward.eq(0)) {
		delete reducedBlock.reward;
	}
	if (reducedBlock.transactions && reducedBlock.transactions.length === 0) {
		delete reducedBlock.transactions;
	}
	/* tslint:enable:no-delete */

	return reducedBlock;
};

export const getId = (blockBytes: Buffer): string => {
	const hashedBlock = hash(blockBytes);
	// tslint:disable-next-line no-magic-numbers
	const temp = Buffer.alloc(8);
	// tslint:disable-next-line no-magic-numbers no-let
	for (let i = 0; i < 8; i += 1) {
		// tslint:disable-next-line no-magic-numbers
		temp[i] = hashedBlock[7 - i];
	}

	const id = BigNum.fromBuffer(temp).toString();

	return id;
};

export const setHeight = (
	block: BlockHeader,
	lastBlock: BlockHeader,
): BlockHeader => {
	block.height = lastBlock.height + 1;

	return block;
};

export const loadMemTables = async (
	storage: Storage,
	tx: StorageTransaction,
): Promise<{
	readonly blocksCount: number;
	readonly genesisBlock: BlockHeaderJSON;
}> => {
	const promises = [
		storage.entities.Block.count({}, {}, tx),
		storage.entities.Block.getOne({ height: 1 }, {}, tx),
	];

	const [blocksCount, genesisBlock] = await tx.batch(promises);

	return {
		blocksCount,
		genesisBlock,
	};
};
