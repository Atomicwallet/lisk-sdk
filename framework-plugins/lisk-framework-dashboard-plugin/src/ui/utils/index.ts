/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { apiClient } from '@liskhq/lisk-client';
import { Block, Transaction } from '../types';

const MAX_BLOCKS = 103 * 3;
const MAX_TRANSACTIONS = 150;

export const updateStatesOnNewBlock = (
	client: apiClient.APIClient,
	newBlockStr: string,
	blocks: Block[],
	confirmedTransactions: Transaction[],
	unconfirmedTransactions: Transaction[],
): {
	blocks: Block[];
	confirmedTransactions: Transaction[];
	unconfirmedTransactions: Transaction[];
} => {
	const newBlock = client.block.toJSON(client.block.decode(newBlockStr));
	const newBlocks = [...blocks, newBlock].slice(-1 * MAX_BLOCKS) as Block[];

	for (const trs of ((newBlock as unknown) as Block).payload) {
		confirmedTransactions.push(trs);
	}
	const confirmedTransactionsIds = confirmedTransactions.map(t => t.id);
	const newUnconfirmedTransactions = unconfirmedTransactions
		.filter(t => !confirmedTransactionsIds.includes(t.id))
		.slice(-1 * MAX_TRANSACTIONS);

	const newConfirmedTransactions = confirmedTransactions.slice(-1 * MAX_TRANSACTIONS);

	return {
		blocks: newBlocks,
		confirmedTransactions: newConfirmedTransactions,
		unconfirmedTransactions: newUnconfirmedTransactions,
	};
};

export const updateStatesOnNewTransaction = (
	client: apiClient.APIClient,
	newTransactionStr: string,
	unconfirmedTransactions: Transaction[],
): Transaction[] => {
	const transaction = client.transaction.toJSON(client.transaction.decode(newTransactionStr));
	return [...unconfirmedTransactions, transaction].slice(-1 * MAX_TRANSACTIONS) as Transaction[];
};
