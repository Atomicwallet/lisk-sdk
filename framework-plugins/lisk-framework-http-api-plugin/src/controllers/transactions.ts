/*
 * Copyright © 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { isBase64String } from '@liskhq/lisk-validator';
import { Request, Response } from 'express';
import { BaseChannel, PluginCodec } from 'lisk-framework';

export const getTransaction = (
	channel: BaseChannel,
	codec: PluginCodec,
) => async (req: Request, res: Response): Promise<void> => {
	const transactionId = req.params.id;

	// 400 - Client Side Error
	if (!transactionId || !isBase64String(transactionId)) {
		res.status(400).send({
			errors: [
				{ message: 'The transaction id parameter should be a base64 string.' },
			],
		});
		return;
	}

	let transaction: string;

	try {
		transaction = await channel.invoke<string>('app:getTransactionByID', {
			id: transactionId,
		});
	} catch (error) {
		if (
			(error as Error).message ===
			'Specified key transactions:id:×m does not exist'
		) {
			// 404 - Not Found Error
			res.status(404).send({
				errors: [
					{ message: `The transaction with id ${transactionId} not found.` },
				],
			});
			return;
		}

		// 500 - Server Error
		throw error;
	}

	// 200 - Response
	res.status(200).json(codec.decodeTransaction(transaction));
};
