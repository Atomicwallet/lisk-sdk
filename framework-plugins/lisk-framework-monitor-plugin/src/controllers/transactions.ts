/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { Request, Response } from 'express';
import { BaseChannel } from 'lisk-framework';
import { PeerInfo, SharedState } from '../types';

const getAverage = (
	transactionsStats: Pick<SharedState['transactions'], 'transactions'>,
): number => {
	let transactionCount = 0;
	let total = 0;

	for (const transactionStats of Object.values(transactionsStats.transactions)) {
		transactionCount += 1;
		total += transactionStats.count;
	}

	return total / transactionCount;
};

export const getTransactionStats = (channel: BaseChannel, state: SharedState) => async (
	_req: Request,
	res: Response,
): Promise<void> => {
	const { transactions } = state;
	res.json({
		data: {
			transactions: transactions.transactions,
			connectedPeers: (await channel.invoke<ReadonlyArray<PeerInfo>>('app:getConnectedPeers'))
				.length,
			averageReceivedTransactions: getAverage(transactions),
		},
		meta: {},
	});
};
