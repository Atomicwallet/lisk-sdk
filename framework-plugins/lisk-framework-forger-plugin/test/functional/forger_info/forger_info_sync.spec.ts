/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { testing } from 'lisk-framework';
import {
	createApplicationEnv,
	closeApplicationEnv,
	getForgerInfoByPublicKey,
	getForgerPlugin,
	waitNBlocks,
	waitTill,
} from '../../utils/application';
import { getRandomAccount } from '../../utils/accounts';
import { createTransferTransaction } from '../../utils/transactions';

describe('Forger Info Sync', () => {
	let appEnv: testing.ApplicationEnv;
	let accountNonce = 0;
	let networkIdentifier: Buffer;

	beforeAll(async () => {
		appEnv = createApplicationEnv('sync');
		await appEnv.startApplication();
		// The test application generates a dynamic genesis block so we need to get the networkID like this
		networkIdentifier = appEnv.networkIdentifier;
	});

	afterAll(async () => {
		await closeApplicationEnv(appEnv);
	});

	it('should sync information from scratch on startup', async () => {
		// Arrange
		let forgerPluginInstance = getForgerPlugin(appEnv.application);
		const account = getRandomAccount();
		const transaction = createTransferTransaction({
			amount: '2',
			recipientAddress: account.address,
			fee: '0.3',
			nonce: accountNonce,
			networkIdentifier,
		});
		accountNonce += 1;
		await appEnv.application['_channel'].invoke('app:postTransaction', {
			transaction: transaction.getBytes().toString('hex'),
		});
		await waitNBlocks(appEnv.application, 1);
		await waitTill(2000);
		const { generatorPublicKey } = appEnv.application['_node']['_chain'].lastBlock.header;
		const forgerInfo = await getForgerInfoByPublicKey(forgerPluginInstance, generatorPublicKey);
		// Make sure forger info is not changed
		expect(forgerInfo).toMatchSnapshot();

		// Act
		// Reset forger info manually
		await forgerPluginInstance['_forgerPluginDB'].clear();

		// Close application
		await closeApplicationEnv(appEnv, { clearDB: false });

		// Start the application again
		await appEnv.startApplication();
		forgerPluginInstance = getForgerPlugin(appEnv.application);

		await waitTill(2000);
		// Get forger info
		const forgerInfoAfterRestart = await getForgerInfoByPublicKey(
			forgerPluginInstance,
			generatorPublicKey,
		);

		// Assert
		expect(forgerInfo).toEqual(forgerInfoAfterRestart);
	});
});
