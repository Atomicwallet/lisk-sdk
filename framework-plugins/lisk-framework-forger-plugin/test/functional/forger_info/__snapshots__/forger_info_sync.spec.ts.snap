// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forger Info Sync should sync information from scratch on startup 1`] = `
Object {
  "totalMissedBlocks": 0,
  "totalProducedBlocks": 1,
  "totalReceivedFees": 140000n,
  "totalReceivedRewards": 0n,
  "votesReceived": Array [],
}
`;
